src/DynamicData/Cache/Internal/TransformAsync.cs

```diff
@@ -3,71 +3,60 @@
-                                                                                 finally
-                                                                                 {
-                                                                                     asyncLock.Release();
-                                                                                 }
-                                                                             }).Concat();
-
-                                                                             transformer = transformer.Synchronize(locker).Merge(forced);
-                                                                         }
-
-                                                                         return transformer.SubscribeSafe(observer);
-                                                                     });
-
-    private async Task<IChangeSet<TDestination, TKey>> DoTransform(ChangeAwareCache<TransformedItemContainer, TKey> cache, Func<TSource, TKey, bool> shouldTransform)
-    {
-        var toTransform = cache.KeyValues.Where(kvp => shouldTransform(kvp.Value.Source, kvp.Key)).Select(kvp => new Change<TSource, TKey>(ChangeReason.Update, kvp.Key, kvp.Value.Source, kvp.Value.Source)).ToArray();
+    public IObservable<IChangeSet<TDestination, TKey>> Run() =>
+        Observable.Create<IChangeSet<TDestination, TKey>>(observer =>
+        {
+            var cache = new ChangeAwareCache<TransformedItemContainer, TKey>();
 
-        var transformed = await Task.WhenAll(toTransform.Select(Transform)).ConfigureAwait(false);
+            var transformer = source.Select(changes => DoTransform(cache, changes)).Concat();
 
-        return ProcessUpdates(cache, transformed);
-    }
+            if (forceTransform is not null)
+            {
+                var locker = new object();
+                var forced = forceTransform.Synchronize(locker)
+                    .Select(shouldTransform => DoTransform(cache, shouldTransform)).Concat();
+
+                transformer = transformer.Synchronize(locker).Merge(forced);
+            }
+
+            return transformer.SubscribeSafe(observer);
+        });
 
-    private async Task<IChangeSet<TDestination, TKey>> DoTransform(ChangeAwareCache<TransformedItemContainer, TKey> cache, IChangeSet<TSource, TKey> changes)
+    private IObservable<IChangeSet<TDestination, TKey>> DoTransform(
+        ChangeAwareCache<TransformedItemContainer, TKey> cache, Func<TSource, TKey, bool> shouldTransform)
     {
-        var transformed = await Task.WhenAll(changes.Select(Transform)).ConfigureAwait(false);
+        var toTransform = cache.KeyValues.Where(kvp => shouldTransform(kvp.Value.Source, kvp.Key)).Select(kvp =>
+            new Change<TSource, TKey>(ChangeReason.Update, kvp.Key, kvp.Value.Source, kvp.Value.Source)).ToArray();
+
+        return toTransform.Select(change => Observable.Defer(() => Transform(change).ToObservable()))
```

Comment:
I think `Observable.FromAsync(() => Transform(change))` gives the same semantics as `Defer` plus `ToObservable` and more clearly expresses your intent.

Reply:
So, you could do:

```csharp
return toTransform.Select(change => Transform(change))
        .Merge()
        .ToArray()
        .Select(transformed => ProcessUpdates(cache, transformed));
```

Or, if you want to provide the maximum concurrency parameter:

```csharp
return toTransform.Select(change => Observable.FromAsync(() => Transform(change)))
        .Merge(maximumConcurrency)
        .ToArray()
        .Select(transformed => ProcessUpdates(cache, transformed));
```

Reply:
Agreed, I believe `Defer()` is redundant here.

Reply:
Defer is necessary, otherwise the task is invoked immediately, which would result in the merge limiter having no effect as it would be the results being limited not the actual tasks.

Reply:
No, `Observable.FromAsync()` already works that way. The async function returning `Task` (I.E. `Transform`) is not invoked until the observable is subscribed to.

Observe:

```cs
var whenCompletedSource = new TaskCompletionSource<int>();

var whenCompleted = Observable.FromAsync(() =>
{
    Console.WriteLine("Observable.FromAsync()");
    return whenCompletedSource.Task;
});

Console.WriteLine("whenCompleted.Subscribe()");
whenCompleted.Subscribe(x => Console.WriteLine($"whenCompleted.OnNext({x})"));

Console.WriteLine("whenCompletedSource.SetResult(7)}");
whenCompletedSource.SetResult(7);
```

Reply:
Your example is for a single invocation. What I'd be interested to see is to try and limit say a 1000 tasks in a way that that only 4 of them would run concurrently

Reply:
#### TL;DR

The function passed to `Observable.FromAsync` gets run once per subscription but not until the subscription happens (like `Defer`) and `Merge(x)` does not subscribe to the X+1-th item until one of the previous streams completes.

You need `Defer` if you invoke a method to get a task and then convert to an observable, but I think `Defer` plus `ToObservable` is semantically the same as `Observable.FromAsync`, so coder's choice, but I do feel like `Observable.FromAsync` is clearer.

#### Code
I used this test... It's only 20, not 1000, but I don't think that changes anything...

```csharp
async Task Main()
{
    // These two seem to be the same
    var obsAsync = Observable.FromAsync(AsyncThing);
    // var obsAsync = Observable.Defer(() => AsyncThing().ToObservable());

    var obs = Observable.Range(0, 20).Select(n => obsAsync.Spy($"Async #{n}"));
    
    // Only 3 AsyncThings at a time
    await obs.Merge(3);
    
    // Let them all fly
    //await obs.Merge();
}

static int current = 0;
static int counter = 0;

async Task<int> AsyncThing()
{
    int n = Interlocked.Increment(ref current);
    Console.WriteLine($"Starting AsyncThing #{n} ({Interlocked.Increment(ref counter)})");
    
    try
    {
        await Task.Delay(100);
        return n;
    }
    finally
    {
        Console.WriteLine($"Ending AsyncThing #{n} ({Interlocked.Decrement(ref counter)})");
    }
}
```

`AsyncThing` is invoked once per subscription.  The Observable is created, but the async method is not invoked until the subscribe.

With `Merge()`, `counter` gets up to 20 because they're all running at once.  With `Merge(3)`, then `counter` never gets bigger than 3.

Reply:
Thanks for looking into this. I add the overload and some testing soon.

Reply:
@JakenVeina I can confirm that Observable,FromAsync() works like the Defer as you suggested.  I was not willing to change what I had until I had a test.  I get similar run times as previously.

![image](https://github.com/reactivemarbles/DynamicData/assets/9879866/82afcdf6-0472-4cc9-97ca-6d6111bdf1b5)

I have a question for you against the test.